libs/utils/src/properties.c

```diff
@@ -103,12 +132,275 @@ static void updateBuffers(char **key, char ** value, char **output, int outputPo
+        return CELIX_SUCCESS; // print error and ignore
+    }
+
+    celix_properties_entry_t* entry =
+        celix_properties_createEntry(properties, strValue, longValue, doubleValue, boolValue, versionValue);
+    if (!entry) {
+        return CELIX_ENOMEM;
+    }
+
+    const char* mapKey = key;
+    if (!celix_stringHashMap_hasKey(properties->map, key)) {
+        // new entry, needs new allocated key;
+        mapKey = celix_properties_createString(properties, key);
+        if (!mapKey) {
+            celix_properties_destroyEntry(properties, entry);
+            return CELIX_ENOMEM;
+        }
+    }
+
+    celix_status_t status = celix_stringHashMap_put(properties->map, mapKey, entry);
+    if (status != CELIX_SUCCESS) {
+        celix_properties_destroyEntry(properties, entry);
+        if (mapKey != key) {
+            celix_properties_freeString(properties, (char*)mapKey);
+        }
+    }
+    return status;
+}
+
+static void celix_properties_removeKeyCallback(void* handle, char* key) {
+    celix_properties_t* properties = handle;
+    celix_properties_freeString(properties, key);
+}
+
+static void celix_properties_removeEntryCallback(void* handle,
+                                                 const char* key __attribute__((unused)),
+                                                 celix_hash_map_value_t val) {
+    celix_properties_t* properties = handle;
+    celix_properties_entry_t* entry = val.ptrValue;
+    celix_properties_destroyEntry(properties, entry);
+}
+
+celix_properties_t* celix_properties_create() {
+    celix_properties_t* props = malloc(sizeof(*props));
+    if (props != NULL) {
+        celix_string_hash_map_create_options_t opts = CELIX_EMPTY_STRING_HASH_MAP_CREATE_OPTIONS;
+        opts.storeKeysWeakly = true;
+        opts.initialCapacity = (unsigned int)ceil(CELIX_SHORT_PROPERTIES_OPTIMIZATION_ENTRIES_SIZE / 0.75);
+        opts.removedCallbackData = props;
+        opts.removedCallback = celix_properties_removeEntryCallback;
+        opts.removedKeyCallback = celix_properties_removeKeyCallback;
+        props->map = celix_stringHashMap_createWithOptions(&opts);
+        props->currentStringBufferIndex = 0;
+        props->currentEntriesBufferIndex = 0;
+        if (props->map == NULL) {
+            free(props);
+            props = NULL;
+        }
+    }
+    return props;
+}
+
+void celix_properties_destroy(celix_properties_t* props) {
+    if (props != NULL) {
+        celix_stringHashMap_destroy(props->map);
+        free(props);
+    }
+}
+
+celix_properties_t* celix_properties_load(const char* filename) {
+    FILE* file = fopen(filename, "r");
+    if (file == NULL) {
+        celix_err_pushf("Cannot open file '%s'", filename);
+        return NULL;
+    }
+    celix_properties_t* props = celix_properties_loadWithStream(file);
+    fclose(file);
+    return props;
+}
+
+static void celix_properties_parseLine(const char* line, celix_properties_t* props) {
```

Comment:
I agree that this can and should be improved, but I propose to do this in a separate PR: #685. 
Again to prevent that the scope of this PR becomes much larger.